magefiles/magefile.go

```diff
@@ -567,8 +655,93 @@ func pwd() string {
+	os.MkdirAll(runtimesDir, 0770)
+
+	clientPath := filepath.Join(home, "porter"+xplat.FileExt())
+	if clientFound, _ := pkg.IsCommandAvailable(clientPath, version, "--version"); !clientFound {
+		log.Println("Porter client not found at", clientPath)
+		log.Println("Installing porter into", home)
+		opts := downloads.DownloadOptions{
+			UrlTemplate: "https://cdn.porter.sh/{{.VERSION}}/porter-{{.GOOS}}-{{.GOARCH}}{{.EXT}}",
+			Name:        "porter",
+			Version:     version,
+			Ext:         xplat.FileExt(),
+		}
+		if runtime.GOOS == "darwin" && runtime.GOARCH == "arm64" {
+			// we don't yet publish arm64 for porter
+			opts.UrlTemplate = "https://cdn.porter.sh/{{.VERSION}}/porter-darwin-amd64"
+		}
+		mgx.Must(downloads.Download(home, opts))
+	}
+
+	runtimePath := filepath.Join(home, "runtimes", "porter-runtime")
+	//if runtimeFound, _ := pkg.IsCommandAvailable(runtimePath, version, "--version"); !runtimeFound {
+	if _, err := os.Stat(runtimePath); err != nil {
+		log.Println("Porter runtime not found at", runtimePath)
+		opts := downloads.DownloadOptions{
+			UrlTemplate: "https://cdn.porter.sh/{{.VERSION}}/porter-linux-amd64",
+			Name:        "porter-runtime",
+			Version:     version,
+		}
+		mgx.Must(downloads.Download(runtimesDir, opts))
+	}
+}
+
 // Run porter using the local storage, not the in-cluster storage
-func porter(args ...string) shx.PreparedCommand {
-	return shx.Command("porter").Args(args...).
-		Env("PORTER_DEFAULT_STORAGE=", "PORTER_DEFAULT_STORAGE_PLUGIN=mongodb-docker")
+func buildPorterCmd(args ...string) shx.PreparedCommand {
+	mg.SerialDeps(porter.UseBinForPorterHome, porter.EnsurePorter)
+	return must.Command(filepath.Join(pwd(), "bin/porter")).Args(args...).
+		Env("PORTER_DEFAULT_STORAGE=",
+			"PORTER_DEFAULT_STORAGE_PLUGIN=mongodb-docker",
+			fmt.Sprintf("PORTER_HOME=%s", filepath.Join(pwd(), "bin")))
+	//mg.Deps(EnsureLocalPorter)
+	// return shx.Command(filepath.Join(GetPorterHome(), "porter")).Args(args...).
+	// 	Env("PORTER_DEFAULT_STORAGE=", "PORTER_DEFAULT_STORAGE_PLUGIN=mongodb-docker")
+}
+
+func BuildLocalPorterAgent() {
+	mg.SerialDeps(porter.UseBinForPorterHome, porter.EnsurePorter)
+	porterRegistry := "ghcr.io/getporter"
+	buildImage := func(img string) error {
+		_, err := shx.Output("docker", "build", "-t", img,
+			"--build-arg", fmt.Sprintf("PORTER_VERSION=%s", porterVersion),
+			"--build-arg", fmt.Sprintf("REGISTRY=%s", porterRegistry),
+			"-f", "tests/integration/testdata/Dockerfile.k8s-plugin-agent", ".")
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	if os.Getenv("PORTER_AGENT_REPOSITORY") != "" && os.Getenv("PORTER_AGENT_VERSION") != "" {
+		localAgentImgName = fmt.Sprintf("%s:%s", os.Getenv("PORTER_AGENT_REPOSITORY"), os.Getenv("PORTER_AGENT_VERSION"))
+	}
+	err := buildImage(localAgentImgName)
+	mgx.Must(err)
+}
+
+//generatedCodeFilter remove generated code files from coverage report
+func generatedCodeFilter(filename string) error {
```

Comment:
I was wondering how you so dramatically increased code coverage in this patch! 😂

Reply:
I hope you don't think it's cheating  :)